tensorflow_quantum/core/ops/tfq_adj_grad_op.cc

```diff
@@ -0,0 +1,403 @@
+    // here slightly.
+    if (max_num_qubits >= 25 || programs.size() == 1) {
+      ComputeLarge(num_qubits, qsim_circuits, maps, partial_fused_circuits,
+                   pauli_sums, gradient_gates, downstream_grads, context,
+                   &output_tensor);
+    } else {
+      ComputeSmall(num_qubits, max_num_qubits, qsim_circuits, maps,
+                   partial_fused_circuits, pauli_sums, gradient_gates,
+                   downstream_grads, context, &output_tensor);
+    }
+    // just to be on the safe side.
+    qsim_circuits.clear();
+    unused_fuse.clear();
+    gate_meta.clear();
+    gradient_gates.clear();
+    num_qubits.clear();
+    maps.clear();
+    pauli_sums.clear();
+    programs.clear();
+    downstream_grads.clear();
+  }
+
+ private:
+  void ComputeSmall(
+      const std::vector<int>& num_qubits, const int max_num_qubits,
+      const std::vector<QsimCircuit>& qsim_circuits,
+      const std::vector<SymbolMap>& maps,
+      const std::vector<std::vector<std::vector<qsim::GateFused<QsimGate>>>>&
+          partial_fused_circuits,
+      const std::vector<std::vector<PauliSum>>& pauli_sums,
+      const std::vector<std::vector<tfq::GradientOfGate>>& gradient_gates,
+      const std::vector<std::vector<float>>& downstream_grads,
+      tensorflow::OpKernelContext* context,
+      tensorflow::TTypes<float, 1>::Matrix* output_tensor) {
+    // Instantiate qsim objects.
+    const auto tfq_for = qsim::SequentialFor(1);
+    using Simulator = qsim::Simulator<const qsim::SequentialFor&>;
+    using StateSpace = Simulator::StateSpace;
+    using State = StateSpace::State;
+
+    auto DoWork = [&](int start, int end) {
+      // Begin simulation.
+      int largest_nq = 1;
+      State sv = StateSpace(largest_nq, tfq_for).CreateState();
+      State scratch = StateSpace(largest_nq, tfq_for).CreateState();
+      State scratch2 = StateSpace(largest_nq, tfq_for).CreateState();
+
+      for (int i = start; i < end; i++) {
+        int nq = num_qubits[i];
+        Simulator sim = Simulator(nq, tfq_for);
+        StateSpace ss = StateSpace(nq, tfq_for);
+        if (nq > largest_nq) {
+          // need to switch to larger statespace.
+          largest_nq = nq;
+          sv = ss.CreateState();
+          scratch = ss.CreateState();
+          scratch2 = ss.CreateState();
+        }
+
+        // (#679) Just ignore empty program
+        if (qsim_circuits[i].gates.size() == 0) {
+          continue;
+        }
+
+        ss.SetStateZero(sv);
+        for (int j = 0; j < partial_fused_circuits[i].size(); j++) {
+          for (int k = 0; k < partial_fused_circuits[i][j].size(); k++) {
+            qsim::ApplyFusedGate(sim, partial_fused_circuits[i][j][k], sv);
+          }
+          if (j == partial_fused_circuits[i].size() - 1) {
+            break;
+          }
+          qsim::ApplyGate(
+              sim, qsim_circuits[i].gates[gradient_gates[i][j].index], sv);
```

Comment:
Could you add a comment? I confused at the first glance of this code.
```
// Apply the original gate of the gradient_gate, not the gradient_gate itself.
```

Reply:
Done.